rpcs3: 0.0.12 -> 0.0.16

vs49688 opened the pull request:
###### Motivation for this change

Updating

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 122140` at 65f95573 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>rpcs3</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/misc/emulators/rpcs3/default.nix:51:5:
```
   |
51 |     license = licenses.gpl2;
   |     ^
```
</li>
  </ul>
</details>

ilian commented:
I wasn't able to derive the GPL license type of RPCS3 so I added an upstream issue for the authors to clarify: rpcs3/rpcs3#10255

ilian commented:
> I just grepped for "any later version", which most of the files seemed to have, hence gpl2Plus.

The only files that matched that pattern are external libraries (e.g. those listed under /Crypto) instead of the source files of the authors.

ilian commented:
> Fair enough, I only changed it because of:
> 
> > Result of `nixpkgs-review pr 122140` at [65f9557](https://github.com/NixOS/nixpkgs/commit/65f955730e97014b08dbdd52f8a7189853a3e8f0) run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> > 1 package built successfully:
> > 1 suggestion:
> 
> I can force-push a new one without the license change for now, if that'd be better.

Sounds good! We can change the license as soon as the linked upstream issue is resolved.

By the way, since we are fetching the source from GitHub, I suggest that we use fetchFromGitHub to fetch the source files:
```
  src = fetchFromGitHub {
    owner = "RPCS3";
    repo = "rpcs3";
    rev = "v${majorVersion}";
    fetchSubmodules = true;
    sha256 = "qceUt3TnzqpK+Ur5DrQ1/g97RJGdynISAcuVJ0KD1Nk=";
  };
```

vs49688 commented:
Done, however I realised that I didn't update the hash so it was still building the older one. It's building on my machine now, but it's not particularly fast...

vs49688 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -25863,7 +25863,7 @@ in
   rootlesskit = callPackage ../tools/virtualization/rootlesskit {};
 
-  rpcs3 = libsForQt514.callPackage ../misc/emulators/rpcs3 { };
+  rpcs3 = libsForQt515.callPackage ../misc/emulators/rpcs3 { };
```
Done

vs49688 commented:
By the way, there's been a response on the license: https://github.com/RPCS3/rpcs3/issues/10255#issuecomment-835512840
Seems the project as a whole is `gpl2Only`, so should we go with that?

Also this https://github.com/RPCS3/rpcs3/issues/10255#issuecomment-835526224 concerns me... Should I switch to master (https://github.com/RPCS3/rpcs3/commit/31b55e0030d35e20f8afd2f27becfbf143d3d169 at time of writing) or stay with the tag?

ilian commented:
> By the way, there's been a response on the license: [RPCS3/rpcs3#10255 (comment)](https://github.com/RPCS3/rpcs3/issues/10255#issuecomment-835512840)
> Seems the project as a whole is `gpl2Only`, so should we go with that?

Sounds good!

> Also this [RPCS3/rpcs3#10255 (comment)](https://github.com/RPCS3/rpcs3/issues/10255#issuecomment-835526224) concerns me... Should I switch to master ([RPCS3/rpcs3@31b55e0](https://github.com/RPCS3/rpcs3/commit/31b55e0030d35e20f8afd2f27becfbf143d3d169) at time of writing) or stay with the tag?

Yes, we should probably switch to master.

SuperSandro2000 commented on pkgs/misc/emulators/rpcs3/default.nix:
```diff
@@ -32,23 +35,27 @@ mkDerivation {
     qtbase qtquickcontrols openal glew vulkan-headers vulkan-loader libpng ffmpeg
-    libevdev python3
-  ] ++ lib.optional pulseaudioSupport libpulseaudio
+    libevdev curl /* wolfssl */ python3
```
```suggestion
    libevdev curl python3
```

SuperSandro2000 commented on pkgs/misc/emulators/rpcs3/default.nix:
```diff
@@ -32,23 +35,27 @@ mkDerivation {
     "-DUSE_SYSTEM_LIBPNG=ON"
     "-DUSE_SYSTEM_FFMPEG=ON"
+    "-DUSE_SYSTEM_CURL=ON"
+    # NB: Can't use this yet, our CMake doesn't include FindWolfSSL.cmake
+    #"-DUSE_SYSTEM_WOLFSSL=ON"
```
Can we patch wolfssl? It should cause minimal rebuilds.

SuperSandro2000 commented on pkgs/misc/emulators/rpcs3/default.nix:
```diff
@@ -1,23 +1,26 @@
+    owner = "RPCS3";
+    repo = "rpcs3";
+    rev = "a4f4b81e6b0c00f4c30f9f5f182e5fe56f9fb03c";
+    fetchSubmodules = true;
```
Please devendor all their sub modules. We do not want to accidentally link against an old curl, ffmpeg and many others.

vs49688 replied:
Not possible without upstream changes. Most of this is handled by the `-DUSE_SYSTEM_*` options (except WolfSSL, obviously), however there's others for which there's no override. The build fails without `fetchSubmodules = true;`

SuperSandro2000 replied:
I am mostly concerned about curl and ffmpeg which should be updated regularly.

vs49688 replied:
Both curl and ffmpeg are de-vendored, see:
 ```nix
    "-DUSE_SYSTEM_FFMPEG=ON"
    "-DUSE_SYSTEM_CURL=ON"
```

vs49688 committed:
rpcs3: 0.0.12-12199-b8477a470 -> 0.0.16-12235-a4f4b81e6

Additionally:
* Use system libcurl, so the patch downloader can work
* Use system zlib
* Use system libusb
* Add note about using system WolfSSL
* Enable SDL2, so the FAudio backend can work
* Change license to gpl2Only, as per [1]

[1]: https://github.com/RPCS3/rpcs3/issues/10255#issuecomment-835512840

vs49688 commented:
I've just de-vendor'd zlib, libusb, and llvm. If that's enough, I'll squash and force-push.

vs49688 commented on pkgs/misc/emulators/rpcs3/default.nix:
```diff
@@ -30,25 +35,32 @@ mkDerivation {
     qtbase qtquickcontrols openal glew vulkan-headers vulkan-loader libpng ffmpeg
-    libevdev python3
-  ] ++ lib.optional pulseaudioSupport libpulseaudio
+    libevdev zlib libusb1 llvm curl python3
```
Gladly, I'll just drop that commit

vs49688 commented:
Ping? I would really like this to be 21.05
